### PR TITLE
Markup <table/> and <td/>

### PR DESCRIPTION
**Pull request by HiromuHota**
## Description of the problems or issues

**Is your pull request related to a problem? Please describe.**

Currently, `<table/>` and `<td/>` are not marked up with hOCR attributes.

**Does your pull request fix any issue.**

N/A

## Description of the proposed changes

This PR adds
- `title="bbox X X X X"` to both of them.
- `class_="ocr_table"` to `<table/>`

## Test plan

- Use hocr-checks to check the conformity to hOCR (this check is actually very basic)
- Check if `table` can be found by class "ocr_table"
- Check if they have bbox

## Checklist

* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [ ] All new and existing tests passed.
* [ ] I have updated the CHANGELOG.md accordingly.

### CODE REVIEW

File: pdftotree/TreeExtract.py [hunk 1]
```diff
@@ -402,42 +402,72 @@ def get_html_others(self, tag: str, box: List[float], page_num: int) -> Element:
...
-    def get_html_table(self, table, page_num) -> Element:
-        table_str = [str(i) for i in table]
+    def get_html_table(self, table: List[float], page_num) -> Optional[Element]:
+        """Recognize a table using tabula and return a DOM element.
+
+        :param table: bbox for a table (top,left,bottom,right)
+        :param page_num: 1-based page number
+        :return: DOM element for a table
+        """
+        self.log.debug(f"Calling tabula at page: {page_num} and area: {table}.")
```
Comment (SenWu) [inline pdftotree/TreeExtract.py:412]: Can we use global logging instead of logging in the class?

File: pdftotree/TreeExtract.py [hunk 2]
```diff
@@ -402,42 +402,72 @@ def get_html_others(self, tag: str, box: List[float], page_num: int) -> Element:
...
-                            text = html.escape(word[0])
-
-                            word_element = self.doc.createElement("span")
-                            col_element.appendChild(word_element)
-                            word_element.setAttribute("class", "ocrx_word")
-                            word_element.setAttribute(
-                                "title", f"bbox {left} {top} {right} {bottom}"
-                            )
-                            word_element.appendChild(self.doc.createTextNode(text))
+        self.log.debug(f"Tabula recognized {len(table_json)} table(s).")
```
Comment (SenWu) [inline pdftotree/TreeExtract.py:416]: See above comment.

File: pdftotree/TreeExtract.py [hunk 3]
```diff
@@ -402,42 +402,72 @@ def get_html_others(self, tag: str, box: List[float], page_num: int) -> Element:
...
+        right = int(table_json[0]["right"])
+        table_element.setAttribute("title", f"bbox {left} {top} {right} {bottom}")
+        for i, row in enumerate(table_json[0]["data"]):
+            row_element = self.doc.createElement("tr")
+            table_element.appendChild(row_element)
+            for j, cell in enumerate(row):
+                # It is not explicitly stated anywhere but tabula seems to use the cell
+                # bbox to represent that of cell itself rather than that of text inside.
+                # Note: bbox could be [0, 0, 0, 0] if tabula recognizes no text inside.
+                box: List[float] = [
```
Comment (SenWu) [inline pdftotree/TreeExtract.py:433]: Can we make this a data class?

File: pdftotree/TreeExtract.py [hunk 4]
```diff
@@ -402,42 +402,72 @@ def get_html_others(self, tag: str, box: List[float], page_num: int) -> Element:
...
+                    cell["left"] + cell["width"],
+                ]
+                cell_element = self.doc.createElement("td")
+                row_element.appendChild(cell_element)
+                elems = get_mentions_within_bbox(box, self.elems[page_num].mentions)
+                if len(elems) == 0:
+                    continue
+                cell_element.setAttribute(
+                    "title",
+                    f"bbox {int(box[1])} {int(box[0])} {int(box[3])} {int(box[2])}",
```
Comment (SenWu) [inline pdftotree/TreeExtract.py:446]: If we have data class for box, then this is much clearer.

File: pdftotree/TreeExtract.py [hunk 5]
```diff
@@ -402,42 +402,72 @@ def get_html_others(self, tag: str, box: List[float], page_num: int) -> Element:
...
+                for elem in elems:
+                    line_element = self.doc.createElement("span")
+                    cell_element.appendChild(line_element)
+                    line_element.setAttribute("class", "ocrx_line")
+                    line_element.setAttribute(
+                        "title",
+                        " ".join(["bbox"] + [str(int(_)) for _ in elem.bbox]),
+                    )
+                    words = self.get_word_boundaries(elem)
+                    for word in words:
```
Comment (SenWu) [inline pdftotree/TreeExtract.py:458]: See the comment above.

**Review by SenWu — APPROVED** [review]
Left few comments.

File: pdftotree/TreeExtract.py (hunk 1 above)
Comment (HiromuHota) [reply pdftotree/TreeExtract.py]: Great idea! Resolved by 2b68c4b9d7aad0013e40d93e9180d2f7d00191ff.

File: pdftotree/TreeExtract.py (hunk 3 above)
Comment (HiromuHota) [reply pdftotree/TreeExtract.py:434]: A data class would be a good idea, but it's only available on Python 3.7 or later.
Instead, I have been actually wondering if we could use [Shapely](https://pypi.org/project/Shapely/) to represent a bbox more broadly in pdftotree.
Shapely comes with R-tree (https://shapely.readthedocs.io/en/latest/manual.html#str-packed-r-tree), which allows fast geometric query like contains or intersect.

Comment (HiromuHota) [reply pdftotree/TreeExtract.py:434]: For now, let's use Shapely for fast geometric query and make an util function that takes `bbox` and returns its string representation for hOCR export.
Once pdftotree drops support for Python 3.6, we can reassess dataclass.

**Review by lukehsiao — APPROVED** [review]

File: pdftotree/TreeExtract.py (hunk 3 above)
Comment (HiromuHota) [reply pdftotree/TreeExtract.py:434]: I'll address this issue in a separate one.

**Pull request merged**